add filters for irrelevant files in ISISDATA

antonhibl opened the pull request:
## Description
added filters for the following files in the ISISDATA download script, these are being imminently removed:

/a_older_versions/*
/kernels/*/former_versions/
/corrupt_files/*
/zzarchive/**
/kernels/*/original
gallileo/kernels/ck/prime_mission//extended_mission/
gallileo/kernels/*/prime_mission/
galileo/kernels/ck/GEM/
clementine1/kernels/ck/save
chandrayaan1/kernels/spk/SAVE_SCS_2017-11-22
/kernels/fk/release./*
cassini/kernels/fk/Archive

## Related Issue
Data storage concerns in ISISDATA area

## How Has This Been Validated?
tested downloadIsisData script on my machine and it works fine with these changes applied

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation change (update to the documentation; no code change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Infrastructure change (changes to things like CI or the build system that do not impact users)

## Checklist:
- [x] I have read and agree to abide by the [Code of Conduct](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/Code-Of-Conduct.md)
- [x] I have read the [**CONTRIBUTING**](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CONTRIBUTING.md) document.
- [x] I have added myself to the [.zenodo.json](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/.zenodo.json) document.

## Licensing
This project is mostly composed of free and unencumbered software released into the public domain, and we are unlikely to accept contributions that are not also released into the public domain. Somewhere near the top of each file should have these words:

> This work is free and unencumbered software released into the public domain. In jurisdictions that recognize copyright laws, the author or authors of this software dedicate any and all copyright interest in the software to the public domain.

- [x] I dedicate any and all copyright interest in this software to the public domain. I make this dedication for the benefit of the public at large and to the detriment of my heirs and successors. I intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

Kelvinrr committed:
added findfeaturesSegment script

antonhibl committed:
adding filters for old files in ISISDATA area

antonhibl committed:
adding script changes and fixing PR

Kelvinrr commented on isis/config/exclude_list.txt:
```diff
@@ -0,0 +1,12 @@
+- clementine1/kernels/ck/save
+- chandrayaan1/kernels/spk/SAVE_SCS_2017-11-22
+- */kernels/fk/release.*/*
+- cassini/kernels/fk/Archive
```
Just to repeat here, might want to put this as a list in the script so users don't have to download another file if they are just wget-ing it.

Kelvinrr commented on isis/scripts/downloadIsisData:
```diff
@@ -110,7 +110,10 @@ def create_rclone_arguments(destination, mission_name, ntransfers=10, rclone_kwa
+    exclude_paths = Path(os.path.dirname(__file__)) / '..' / 'config' / 'exclude_list.txt'
+
+
+    extra_args=[f"{mission_name}",f"{destination}", "--progress", f"--checkers={ntransfers}", f"--transfers={ntransfers}", "--track-renames", f"--log-level={log.getLevelName(log.getLogger().getEffectiveLevel())}", f"--exclude-from={exclude_paths}"]
```
What would happen if the user inputs their own exclude list?

antonhibl committed:
used exlcude instead of exclude-from

antonhibl committed:
cleaned up files and script

antonhibl committed:
added test to see if filtering on rclone ISISDATA worked

antonhibl committed:
modified tests

antonhibl committed:
updating imports in tests

antonhibl committed:
added complete test for filters, list of filter files can be changed

antonhibl committed:
updated test to check args rather than files

antonhibl committed:
removed absolute paths

Kelvinrr commented on isis/pytests/test_IsisDownload.py:
```diff
@@ -59,3 +60,33 @@ def test_create_rclone_args():
+
+
+def test_file_filtering():
+    files_to_exclude = [
+        "/a_older_versions/*",
+        "/kernels/*/former_versions/",
+        "/corrupt_files/*",
+        "/zzarchive/**",
+        "/kernels/*/original",
+        "gallileo/kernels/ck/prime_mission/*/extended_mission/*",
+        "gallileo/kernels/*/prime_mission/",
+        "galileo/kernels/ck/GEM/",
+        "clementine1/kernels/ck/save",
+        "chandrayaan1/kernels/spk/SAVE_SCS_2017-11-22",
+        "*/kernels/fk/release.*/*",
+        "cassini/kernels/fk/Archive"
+    ]
```
If you are sourcing the script, you can probably use the same exclude list from the actual script rather than duplicating here.

Kelvinrr commented on isis/scripts/downloadIsisData:
```diff
@@ -110,7 +113,7 @@ def create_rclone_arguments(destination, mission_name, ntransfers=10, rclone_kwa
         destination = os.path.join(destination, "kernels")
 
-    extra_args=[f"{mission_name}",f"{destination}", "--progress", f"--checkers={ntransfers}", f"--transfers={ntransfers}", "--track-renames", f"--log-level={log.getLevelName(log.getLogger().getEffectiveLevel())}"]
+    extra_args=[f"{mission_name}",f"{destination}", "--progress", f"--checkers={ntransfers}", f"--transfers={ntransfers}", "--track-renames", f"--log-level={log.getLevelName(log.getLogger().getEffectiveLevel())}", f"--exclude={exclude_string}"]
```
You might want to check for additional --include or --filter flags passed into by the user and merge them since rclone docs say mixing them can cause weird behavior. Maybe default to --filter, and grab any user --include and --exclude flags and merge them into the filter flag?

Kelvinrr reviewed (commented):
test looks good 👍🏽 just one major hurdle left.

antonhibl committed:
updated test to use exclude list passed in

antonhibl committed:
Updated filter args

I added the following changes to create_rclone_arguments:

- I added a default filter flag --filter with the value of exclude_string

- I added a check for any additional --include and --exclude flags that may have
  been passed in by the user.

- I added logic to merge any additional --include and --exclude flags with the
  default --filter flag. The include and exclude patterns are concatenated with
  the default exclude_string separated by comma.

- I removed the f"--exclude={exclude_string}" from extra_args list, since the
  default exclude_string is now included in the filter flag

- I added the filter flag to the extra_args list using
  extra_args.extend(filter_args)

All of the above changes ensures that the --include and --exclude flags passed
    in by the user are taken into account while creating extra_args, and also
    the logic will merge these flags with the default filter flag, which is the
    recommended way as per the rclone docs.

antonhibl committed:
accidentally deleted a line

Kelvinrr commented on isis/scripts/downloadIsisData:
```diff
@@ -11,9 +11,12 @@ import tempfile
+from pathlib import Path
+
+exclude_list = [    '/a_older_versions/*',    '/kernels/*/former_versions/',    '/corrupt_files/*',    '/zzarchive/**',    '/kernels/*/original',    'gallileo/kernels/ck/prime_mission/*/extended_mission/*',    'gallileo/kernels/*/prime_mission/',    'galileo/kernels/ck/GEM/',    'clementine1/kernels/ck/save',    'chandrayaan1/kernels/spk/SAVE_SCS_2017-11-22',    '*/kernels/fk/release.*/*',    'ca ...
+exclude_string = "{" + ",".join(exclude_list) + "}"
```
Would this actually filter them out? From the docs they need a + or - sign in front to explicitly include or exclude from the files as --filter is basically a generic version of include and exclude.

Kelvinrr commented on isis/scripts/downloadIsisData:
```diff
@@ -110,8 +113,21 @@ def create_rclone_arguments(destination, mission_name, ntransfers=10, rclone_kwa
+
+    # Check for additional include and exclude flags
+    include_args = [arg for arg in rclone_kwargs if arg.startswith("--include")]
+    exclude_args = [arg for arg in rclone_kwargs if arg.startswith("--exclude")]
```
Would this be robust enough? So capture `--include "stuff"`, `--include="stuff"`, and (I think this is valid too) `--include"stuff"`.

Kelvinrr commented on isis/scripts/downloadIsisData:
```diff
@@ -110,8 +113,21 @@ def create_rclone_arguments(destination, mission_name, ntransfers=10, rclone_kwa
+    exclude_args = [arg for arg in rclone_kwargs if arg.startswith("--exclude")]
 
+    # Merge any additional include and exclude flags with the default filter flag
+    if include_args:
+        filter_args[1] += "," + ",".join([arg.split("=")[1] for arg in include_args])
+    if exclude_args:
+        filter_args[1] += "," + ",".join([arg.split("=")[1] for arg in exclude_args])
```
From the docs: https://rclone.org/filtering/

`--filter +` doesn't not import exclude `*` like include does. If the user specifies --include. If there is an --include flag, we might want to add `- *` at the bottom to simulate this? Remember, the goal is to honor the user's intention when they input their own flags, if they explicitly want the stuff we are filtering, we should be able to make that work naturally. 

Also, this doesn't seem to be accounting for the "-" or "+" in --filter to specify if it's including or excluding something? I would say read those docs I linked to get an idea of how to make this work.

antonhibl committed:
Added more logic for when the user provides their own filters

along with a few other changes:

   + I added a check for any --filter flag provided by the user, if present it
    will use it and ignore the default filter flag. Otherwise, it will use the default
    filter flag. This is done to take into account if the user has provided any
    specific filter flag, and it will honor the user's intention of providing
    the filter flag.

   + I added a check for any additional --include and --exclude flags passed in
    by the user and merge them with the filter flag. This is to take into
    account any specific include/exclude patterns that the user wants to apply,
    and merge them with the default filter flag.

   + I added a "+" at the end of the filter string if the user has specified an
    --include flag. This simulates the behavior of --include where it includes
    any patterns specified and excludes everything else.

   + I also added a check for filter_args_provided and if provided, it will use
    this flag, else it will use the default filter flag and merge any additional
    include/exclude flags to it.

There is also a test in the pytest file to check if the filter logic works as
expected, run using `pytest`.

antonhibl committed:
fixed tests and adjusted with new filters

antonhibl committed:
Added several tests to pytests

**The first test that was added is `test_rclone_with_auth`**
    This test is designed to check the behavior of the `rclone` function when it
    is called with an `auth` parameter. This test checks that the `rclone`
    function properly passes the "auth" parameter to the underlying subprocess
    call.

 **The second test that was added is `test_create_rclone_args_with_no_kwargs`**
    This test is designed to check the behavior of the `create_rclone_args`
    function when it is called with no keyword arguments. This test checks that
    the `create_rclone_args` function properly handles the case when it is
    called with no keyword arguments and returns the correct list of arguments
    to be passed to the underlying subprocess call

 **The third test that was added is `test_file_filtering_with_hidden_files`**
    This test is designed to check the behavior of the `file_filtering` function
    when it is called with hidden files in the specified directory. This test
    checks that the `file_filtering` function properly filters out hidden files
    and only returns the non-hidden files in the specified directory.

I have tested to confirm these all run and pass on my machine.

antonhibl committed:
added more tests to pytest

**added `test_rclone` test**
    This test mocks the `subprocess.Popen` function and checks that the output
    of the rclone function matches the expected output when the function is
    invoked with the arguments `lsf`, `test`, `["-l", "-R", "--format", "p",
    "--files-only"]`, `True`, and `True`.

**added `test_rclone_unknown_exception` test**
    This test mocks the `subprocess.Popen` function and checks that the `rclone`
    function raises an exception when an unknown exception is encountered. This
    test uses a mocked class that raises an exception when it is initialized.

I have tested and confirmed these to work on my system.

Kelvinrr commented on isis/scripts/downloadIsisData:
```diff
@@ -110,8 +113,26 @@ def create_rclone_arguments(destination, mission_name, ntransfers=10, rclone_kwa
+        filter_args = filter_args_provided
+    else:
+        if include_args:
+            filter_args[1] += "," + ",".join([arg.split("=")[1] for arg in include_args]) + "+"
+        if exclude_args:
+            filter_args[1] += "," + ",".join([arg.split("=")[1] for arg in exclude_args])
```
Looking at this, it still looks like the flag isn't being set correctly. remember, from the docs I linked, that regexes that need to be includes need to be prefixed with "+" and things that are to be ignored with "-" for every regex. These basically have to be added as there own --filter flag. So you can't use the "{-x,- y, +z }" syntax. e.g `rclone --filter "- gallileo/kernels/ck/prime_mission/*/extended_mission/*" --filter "- /a_older_versions/* ..."`

also, remember if the user passes in a --include, you need to append `--filter "- *"`, in order to replicate the functionality of --include

antonhibl committed:
fixed filter args

    fixed how filters are input, patterns still aren't working I believe I need
    to look at their patterns and ensure the adhere to rclone documentation
    syntactical instructions.

antonhibl committed:
updated filter list,it starts search from {mission_name}/kernels/

antonhibl committed:
finally fixed filters and tests

Kelvinrr commented:
tested this locally and I am still getting some of the kernels that are supposed to be filtered

antonhibl commented:
@Kelvinrr shoot, what files were you seeing not filtering? I was testing by downloading the galileo area and it seemed to work for those files for me; specifically galileo/kernels/ck/GEM/ was filtered when I tested on my machine.

Kelvinrr commented:
It should get everything under extended mission and those things were still being included when downloading Galileo.

Kelvinrr commented on isis/scripts/downloadIsisData:
```diff
@@ -11,9 +11,59 @@ import tempfile
+        'corrupt_files/*',
+        'zzarchive/**',
+        '*/original/',
+        'ck/prime_mission/*/extended_mission/*',
```
I think this is the problem, there are two filters mashed into one. Also the asterisks seem to be messing up the filters. I guess rclone doesn't wants either double asterisks or a substring. 


This seemed to be working for me. Added a few things on top of the existing list. 

```python 
exclude_list = [
        'a_older_versions/',
        'former_versions/',
        'corrupt_files/',
        'zzarchive/',
        'original/',
        'prime_mission/',
        'extended_mission/',
        'prime_mission/',
        'GEM/',
        'save/',
        'SAVE_SCS_2017-11-22/',
        'release*/',
        'Archive/',
        'ek/',
        '*.lbl',
        'unitVectors/',
        '*.pdf'
    ]
```

Kelvinrr commented:
testing filter also failed: 

```shell
(isisdev23) IGSWZAWGLT00017:scripts krodriguez$ ./downloadIsisData galileo /tmp/isisdata -vv -n5 --include "**/ck/*"
Traceback (most recent call last):
  File "/Users/krodriguez/repos/isis/isis/scripts/./downloadIsisData", line 51, in <module>
    include_argers = [arg for arg in sys.argv if arg.startswith("--include")][0].split("=")[1]
IndexError: list index out of range
```

as I said in a previous comment, user can input `--include "filter*` or `--include="filter"`. Why parse the arg yourself when argparse can do it for you?

Kelvinrr committed:
added stuff

Kelvinrr committed:
fixed parsing issues

Kelvinrr committed:
cleaning up

Kelvinrr committed:
more cleaning up

Kelvinrr commented:
I think this is good, but we should wait for #5111 to merge before we merge this so we can get good tests for any changes to the download script using that new application.

Kelvinrr reviewed (dismissed):
Im just gonna approve this so we can shrink the data area already.

antonhibl committed:
Merge branch 'dev' into dev

antonhibl committed:
Merge branch 'dev' into dev

antonhibl committed:
Merge branch 'DOI-USGS:dev' into dev

antonhibl committed:
Merge branch 'DOI-USGS:dev' into dev

antonhibl committed:
Merge branch 'DOI-USGS:dev' into dev

AustinSanders commented:
@Kelvinrr @antonhibl Is this ready to be merged?

antonhibl commented:
@AustinSanders I think so but would double check with Kelvin as he was still working on this afaik @Kelvinrr

antonhibl committed:
added new regex filter

antonhibl committed:
re-implementing kelvin's changes.

    accidentally re-based and merged over them.

antonhibl committed:
Merge branch 'DOI-USGS:dev' into dev

antonhibl committed:
comma typo

antonhibl committed:
fixed capitalization.

antonhibl committed:
fixed i.e. shortened the regex pattern.

    shortened it to spk/spk_psp_rec* as the paths it searches in mission areas
    are actually sub folders of the mission folder. I believe this method saves
    time and thus it is important to preserve behavior here.

AustinSanders committed:
Add dry run flag

antonhibl commented:
In the future adding the ability to have mission specific exclude/include patterns and lists be applied depending what base area is specified would be very useful as it would allow for more granularity in filtering and possibly better mission by mission optimizations and speed-ups.

AustinSanders committed:
Roll back change to dry run
